### PR TITLE
chore: additional test cases for TestCalcJoinPoolShares (multi-coin swap)

### PR DESCRIPTION
**Pull request by alexanderbez**
Add two additional test cases to `TestCalcJoinPoolShares` where `tokensIn` contains multiple coins where a swap is performed.

### CODE REVIEW

File: x/gamm/pool-models/balancer/amm_test.go [hunk 1]
```diff
@@ -361,6 +361,60 @@ func TestCalcJoinPoolShares(t *testing.T) {
...
+			},
+			tokensIn: sdk.NewCoins(
+				sdk.NewInt64Coin("uosmo", 25_000),
+				sdk.NewInt64Coin("uatom", 25_000),
+			),
+			expectErr: false,
+			// TODO: Verify expectShares value using manual calculation using the
+			// following as reference:
+			// https://github.com/osmosis-labs/osmosis/blob/6125f81d1ad95c73e03d86c3d5e96ceef3f8c12f/x/gamm/pool-models/balancer/amm_test.go#L348-L357
+			expectShares: sdk.NewInt(2500000000000),
```
Comment (alexanderbez) [inline x/gamm/pool-models/balancer/amm_test.go:412]: I'd expect the shares to be less compared to the zero-fee case above???

Comment (stackman27) [reply x/gamm/pool-models/balancer/amm_test.go:412]: Here's what i got 
`1249999992187.5000976562484741211204528803586959937214` 

[Wolfram](https://www.wolframalpha.com/input?i=100000000000000000000*%28%281+%2B+%2825000%2F1000000000000%29%29%5E0.5+-+1%29)

File: x/gamm/pool-models/balancer/amm_test.go [hunk 2]
```diff
@@ -361,6 +361,60 @@ func TestCalcJoinPoolShares(t *testing.T) {
...
```
Comment (alexanderbez) [inline x/gamm/pool-models/balancer/amm_test.go:385]: @stackman27 can you verify & confirm `2500000000000` is correct here (zero fee)?

File: x/gamm/pool-models/balancer/amm_test.go (hunk 1 above)
Comment (alexanderbez) [reply x/gamm/pool-models/balancer/amm_test.go:412]: That's quite the difference. Are we sure the formula takens into account multiple assets?

Comment (ValarDragon) [reply x/gamm/pool-models/balancer/amm_test.go:412]: I'm confused, the shares should be the same

This is a join pool at perfect ratio, so there has been no swap that occured

Comment (stackman27) [reply x/gamm/pool-models/balancer/amm_test.go:412]: So if the swap actually occurred, then would the expectShares just be the (SharesOut + SwapFee)?

Comment (alexanderbez) [reply x/gamm/pool-models/balancer/amm_test.go:412]: You're right. Should we change the amounts?

**Comment by stackman27** [comment]
Can we close this as it's replaced by #1732?

**Comment by alexanderbez** [comment]
> Can we close this as it's replaced by #1732?

I would prefer that it'd be based off of `bez/TestCalcJoinPoolShares-swap-tests` and merged into that, but I'm open to closing mine as well.

File: x/gamm/pool-models/balancer/amm_test.go (hunk 1 above)
Comment (ValarDragon) [reply x/gamm/pool-models/balancer/amm_test.go:420]: rather, I was confused by the confusion -- it looks correct to me

Comment (stackman27) [reply x/gamm/pool-models/balancer/amm_test.go:420]: @ValarDragon how would i test for situations where maybe swap has occurred and we take fees? I'm trying to test it and this is what i get;
expectedShares: 2498749980000, actualShares: 2500000000000
```
{
			// Expected output from Balancer paper (https://balancer.fi/whitepaper.pdf) using equation (25) on page 10:
			// P_issued = P_supply * ((1 + (A_t * swapFeeRatio  / B_t))^W_t - 1)
			//
			// 1_249_374_990_000 = 100 * 10^18 * (( 1 + (25,000 * (1 - (1 - 0.5) * 0.001) / 1_000_000_000_000))^0.5 - 1)
			//
			// where:
			// 	P_supply = initial pool supply = 100 * 10^18 (set at pool creation, same for all new pools)
			//	A_t = amount of deposited asset = 50,000
			//	B_t = existing balance of deposited asset in the pool prior to deposit = 1,000,000,000,000
			//	W_t = normalized weight of deposited asset in pool = 0.5 (equally weighted two-asset pool)
			// 	swapFeeRatio = (1 - (1 - W_t) * swapFee)
			// Plugging all of this in, we get:
			// 	Full solution: https://www.wolframalpha.com/input?i=100+*10%5E18*%28%281+%2B+%2850000*%281+-+%281-0.5%29+*+0.01%29%2F1000000000000%29%29%5E0.5+-+1%29
			// 	Simplified:  P_issued = 1_249_374_990_000

			name:    "Multi-tokens In: swap equal weights with 0.001 swap fee",
			swapFee: sdk.MustNewDecFromStr("0.001"),
			poolAssets: []balancer.PoolAsset{
				{
					Token:  sdk.NewInt64Coin("uosmo", 1_000_000_000_000),
					Weight: sdk.NewInt(100),
				},
				{
					Token:  sdk.NewInt64Coin("uatom", 1_000_000_000_000),
					Weight: sdk.NewInt(100),
				},
			},
			tokensIn: sdk.NewCoins(
				sdk.NewInt64Coin("uosmo", 25_000),
				sdk.NewInt64Coin("uatom", 25_000),
			),

			expectShares: sdk.NewInt(1249374990000 * 2),
			expectLiq: sdk.NewCoins(
				sdk.NewInt64Coin("uosmo", 25_000),
				sdk.NewInt64Coin("uatom", 25_000),
			),
		},
```

Comment (ValarDragon) [reply x/gamm/pool-models/balancer/amm_test.go:420]: The input is at a perfect ratio, no swap is occurring!

File: x/gamm/pool-models/balancer/amm_test.go [hunk 3]
```diff
@@ -369,6 +369,60 @@ func TestCalcJoinPoolShares(t *testing.T) {
...
```
Comment (ValarDragon) [inline x/gamm/pool-models/balancer/amm_test.go:393]:
```suggestion
// Raises liquidity perfectly by 25_000 / 1_000_000_000_000. Initial number of pool shares = 100 * 10**18 = 10**20
// Expected increase = liquidity_increase_ratio * initial number of pool shares = (25_000 / 1_000_000_000_000) * 10**20 = 2500000000000.0 = 2.5 * 10**12
			expectShares: sdk.NewInt(2.5e12),
```

**Comment by ValarDragon** [comment]
go for it imo

**Comment by mattverse** [comment]
We basically need more test cases for pool with different ratio, plan is to get this merged and get it done in a separate PR

**Review by mattverse — APPROVED** [review]
LGTM

We need test cases for pools with different ratio but this is tbd in a different pr

**Review by ValarDragon — APPROVED** [review]

**Pull request merged**